db/db_versioned.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2024 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package db
+
+import (
+	"context"
+)
+
+type (
+	// KvVersioned is a versioned key-value store, where each key can (but doesn't
+	// have to) have multiple versions of value (corresponding to different heights
+	// in a blockchain)
+	//
+	// A namespace/bucket is considered versioned by default, user needs to use the
+	// NonversionedNamespaceOption() to specify non-versioned namespaces at the time
+	// of creating the versioned key-value store
+	//
+	// Versioning is achieved by using (key + 8-byte version) as the actual storage
+	// key in the underlying DB. For buckets containing versioned keys, a metadata
+	// is stored at the special key = []byte{0}. The metadata specifies the bucket's
+	// name and the key length
+	//
+	// For each versioned key, the special location = key + []byte{0} stores the
+	// key's metadata, which includes the following info:
+	// 1. the version when the key is first created
+	// 2. the version when the key is lastly written
+	// 3. the version when the key is deleted
+	// 4. hash of the key's last written value
+	// If the location does not store a value, the key has never been written.
+	//
+	// Here's an example of a versioned DB which has 3 buckets:
+	// 1. "mta" -- regular bucket storing metadata, key is not versioned
+	// 2. "act" -- versioned namespace, key length = 20
+	// 3. "stk" -- versioned namespace, key length = 8
+	KvVersioned interface {
```

Comment:
I prefer the following interface:
```go
KvVersioned interface{
    Version(ns string, key []byte) (ver uint64, err error)
    WithVersion(ver uint64) KVStore
    CreateNamespace(ns string, keySize uint32) error
}
```
Simplify the definition of `KvVersioned`, it does not support operations such as put/get itself, and must specify the version before proceeding. If need an interface that supports both versioned and non-versioned operations, can define another one.

Reply:
If this is an interface that supports both versioned and non-versioned operations at the same time, a method `IsVersionedNamespaces(ns string)` should also be added, so that when the interface is used, it is known whether to call `Put()` or `SetVersion().Put()`.

Reply:
good suggestions, updated PR
yes, `Put()` is a common interface for both versioned and non-versioned, the code will check inside and process accordingly